### PR TITLE
Make Zephyr build compatible with optional assert()

### PR DESCRIPTION
**Pull request by marc-hb**
2 commits

----

Break down buggy assert(!memcpy_s(...)) into two statements

assert() semantics are very clear: because asserts are optional, the
argument should never have side-effect, period. assert arguments should
be treated exactly the same as logging arguments. So
`assert(!memcpy_s())` is an obvious bug. Use variable to break that down
into two statements.

Fixes commit 0b7fd030b391 ("demux: return config on COMP_CMD_GET_DATA")

Fixes commit 385586a5f664 ("zephyr: implement IDC using P4WQ")

----

Make Zephyr build compatible with optional assert() with __unused

For some legacy reason(s), SOF assert() is not optional. This is unlike
any other assert() anywhere else. This hasn't been a problem yet because
until now SOF assert() has been mapped to a non-optional and limited
Zephyr k_oops()

This commit makes it possible to temporarily or permanently switch the
Zephyr build to Zephyr's __ASSERT(expr, #expr) macro which logs the
asserted expression. This commit does NOT make such a switch.

Note this commit is not enough to make assert() optional in the XTOS
build.

### CODE REVIEW

**Comment by marc-hb** [comment]
Simpler alternative to
- #5756

File: src/schedule/zephyr.c
Comment (andyross) [inline src/schedule/zephyr.c:69]: Totally a style thing, but I'd view `if (memcpy(...)) { assert(false); }` as more readable than creating a variable you need to annotate with __unused because of a compiler warning.  Or even adding a `checked_memcpy()` variant that will implicitly assert on an error?  Meh, probably not worth the effort.

**Review by andyross — APPROVED** [review]
I guess I can +1 a SOF patch if I was talking about the issue vie a sideband. :)

File: src/schedule/zephyr.c
Comment (marc-hb) [reply src/schedule/zephyr.c:69]: Thanks, I'd love to get rid of the `__unused` and this looks like a great way to do that but I think some `assert(some_readable_condition)` macros are able to print `some_readable_condition` somewhere sometimes when they fail depending on which debugguer/trace/crashdump you use... Is it the case with Zephyr's or SOF's assert? If neither then I totally want to switch to your suggestion as is. If at least one of them does then I'm not sure what's best.

Comment (andyross) [reply src/schedule/zephyr.c:69]: Indeed, Zephyr's __ASSERT() will print the text of the expression, and it even has a printf-style format string and arguments you can use.  And this would defeat that, you'd just get the file/line info.  Which is frankly all I want 94% of the time.

I guess mostly I don't actually want to see the assertions when I'm reading code. :)

Comment (marc-hb) [reply src/schedule/zephyr.c:69]: Change of mind: I found that:
- `__unused` is used in the official Zephyr documentation https://docs.zephyrproject.org/apidoc/latest/group__log__api.html#details
- `#define assert(expr) __ASSERT(expr, #expr)` prints the asserted expression, which is nice.

So I went one step further and added one commit that makes the Zephyr build compatible with Zephyr's `__ASSERT` (without actually switching)

File: src/audio/eq_fir/eq_fir.c
Comment (marc-hb) [inline src/audio/eq_fir/eq_fir.c:470]: @singalsu should these be changed to unconditional `panic`s?

File: src/audio/eq_iir/eq_iir.c
Comment (lyakh) [inline src/audio/eq_iir/eq_iir.c:791]: what is this semicolon here for?

Comment (kv2019i) [reply src/audio/eq_iir/eq_iir.c:791]: the extra semicolon indeed looks a bit whacky

**Review by kv2019i — COMMENTED** [review]
@marc-hb I actually started with this approach, but then decided to go the other path in https://github.com/thesofproject/sof/pull/5756 . 

I do agree with @plbossart 's notion (in that PR) that asserts in general are rarely well used and we really should get rid of them. There is a case to be made that in very resource constrained runtime (like SOF), eloborate error handling code may not be worth the code size it takes, and a simple panic on error, is a practical error handling approach. But I think we should not use the "assert()" interface for this, but something more project specific and explicit that forces developer to think. I.e. is this an error that should be handled runtime, or is this a case where practically error recovery is too complex and a simple panic is more appropriate. When parsing IPC commands sent from host, obviously former should be used (FW should not panic if wrong IPC is sent). If the RTOS scheduler state goes undefined, panic is more appropriate. 

So I imagined the verify_assert() to be a first step allowing to incrementally go through the assert uses, and slowly get rid of asserts.

But yeah, your PR is not really in conflict this approach and verify_assert() might not be the best new interface at all. Maybe we should instead go directly to Zephyr interface and convert the asserts into k_panic()'s and __ASSERT()'s.

So ok to approve this, but let's discuss the case of the weird extra semicon before merging.

File: src/audio/eq_fir/eq_fir.c
Comment (singalsu) [reply src/audio/eq_fir/eq_fir.c:470]: Could this fail without panic by with -EINVAL returned if eq_fir_func is NULL? Comp_set_state() seems to return that with errors.

**Comment by marc-hb** [comment]
Please prove me wrong but I think the semantics of `assert` (like them or not) are standard/universal and well defined:
A. `assert()` **interrupts** the program and shares information about why and where
B. `assert()` is an **optional** debugging aid, it's meant to be enabled only for unthinkable errors that can happen only during development, never in stable release branches.

I think Zephyr (and many other projects) define `assert()` like this. In SOF, it looks like `assert()` was abused in some places and should not have been used. 

>  (FW should not panic if wrong IPC is sent). 

Indeed not because in that case BOTH A. and B. are wrong!

A. a wrong IPC should NOT stop the program
B. a wrong IPC can happen in production and handling it should not be an optional debugging aid.

In a situation like this an `assert()` is very obviously wrong, it should be replaced with an error message and error code.

In other places then maybe a non-optional `panic()` is more appropriate.

I think this PR should not be merged before any inappropriate use of `assert()` that it highlights  are replaced by something else. As SOF becomes a Zephyr project, its non-standard use of industry standards should be fixed and made standard and this PR may "hide" some non-standard uses of `assert`; hence my `eq_iir.c` question to @singalsu above. Non standard practices were "merely" confusing when SOF stood alone but they can become at best unbearable and at worst incompatible when combined with a more regular codebase.

In any case I think a new macro is neither needed nor a good idea. I mean I don't see what more is needed beyond the standard `assert`, `panic` and error messages and if anyone was confused by the non-standard use of `assert()` in SOF then I think any new and non-standard macro would add to the confusion.

**Comment by andyross** [comment]
(Not getting involved in the bikeshed about what an assertion API should look like, heh, just chiming in on the Zepherese question.)

> Given Zephyr doesn't provide a "BUG_ON()" like interface, maybe you are right and this would just confuse more.

We don't have an equivalent to the Linux "BUG" notion, which terminates the calling thread and unwinds its kernel stack (i.e. leaving the system "up", but likely with a drivers in weird states due to the aborted kernel state).  Really that makes sense only for interactive systems where it's plausible for an operator to get in and do some diagnostics (if nothing else, by making it possible to extract the kernel logs).

But a "#define PANIC_ON(expr) if(expr) { k_panic(); }" (or a slightly fancier one with some logging integration, etc...) wouldn't be a bad thing to have in the framework as an intermediate technique between full app-managed error handling and assertions.

**Comment by marc-hb** [comment]
FWIW, below is one of the ways I locally and temporarily tested this PR and this successfully printed something like `failed params == 0` in the Zephyr logs:

```diff
--- a/src/drivers/intel/hda/hda.c
+++ b/src/drivers/intel/hda/hda.c
@@ -41,6 +41,10 @@ static int hda_set_config(struct dai *dai,  struct ipc_config_dai *common_config
        dai_info(dai, "hda_set_config(): channels %u rate %u", params->channels,
                 params->rate);
 
+       // Sample assert that can be triggered on first aplay on UP2,
+       // long after FW boot.
+       assert(params == 0);
+
        if (params->channels)
                hda->params.channels = params->channels;
        if (params->rate)
```

In `sof/src/include/sof/debug/panic.h`:
```C
#ifdef __ZEPHYR__
#define __ASSERT_ON 1
#include "zephyr/sys/__assert.h"
#define assert(x)  __ASSERT(x, #x)

```

**Comment by kv2019i** [comment]
@marc-hb Can we sort out the semicolon question and proceed with this PR? There is immediate benefit in getting the side-effects fixed. Even if product builds continue to keep asserts on, there's definite value in being able to easily create a binary with the asserts left out. For example, this allows to verify and track the true cost of the asserts in any given platform and use-case.

@andyross Thanks for the Zephyr background info. In SOF, the operational environment is in practise similar to Linux. E.g. we have methods to dump the FW state to SRAM which the host OS (now typically Linux) can grab it and the info can be shown to end-user and attached to a bug report (i.e. all the same methods there are to utilize "BUG_ON()" output). So our case is indeed a bit different than a headless embedded device running Zephyr.

File: src/schedule/zephyr.c
Comment (lgirdwood) [reply src/schedule/zephyr.c:69]: is there any reason why we cant use the native zephyr ```ASSERT()``` and redefine if for xtos usage ?

**Review by lgirdwood — COMMENTED** [review]
@andyross btw, does Zephyr have any equivalent of memcpy_s ?

File: src/schedule/zephyr.c
Comment (marc-hb) [reply src/schedule/zephyr.c:69]: > is there any reason why we cant use the native zephyr ASSERT() and redefine if for xtos usage ?

That should be possible but there are other asserts elsewhere (in non-Zephyr code) that this PR does not touch yet and it is already pretty big.

File: src/audio/eq_fir/eq_fir.c
Comment (singalsu) [reply src/audio/eq_fir/eq_fir.c:470]: Sure I will make it tomorrow. Thanks for spotting these @marc-hb !

**Comment by kv2019i** [comment]
@lgirdwood wrote:
> is there any reason why we cant use the native zephyr ASSERT() and redefine if for xtos usage ?

Zephyr codebase also uses "assert()" (dozens of instances across different drivers). The "ASSERT()" and friends provides more options (and I see no reason why we couldn't use it, but alas first we need to fix the existing assertions to comply with the semantics, i.e. this PR).

File: src/audio/eq_iir/eq_iir.c
Comment (lyakh) [reply src/audio/eq_iir/eq_iir.c:791]: mmh, don't understand, sorry. That warning appears when `assert()` is expanded to nothing, right? So, without that additional semicolon you'd have a single semicolon there. With the additional one you have two of them. Not sure it improves anything ;-) Was it the compiler complaining? I think it isn't very rare to have code like
```
while (do_something())
	;
```
with an obvious empty loop body. Maybe you shouldn't add those superfluous braces?

**Comment by lgirdwood** [comment]
> @lgirdwood wrote:
> 
> > is there any reason why we cant use the native zephyr ASSERT() and redefine if for xtos usage ?
> 
> Zephyr codebase also uses "assert()" (dozens of instances across different drivers). The "ASSERT()" and friends provides more options (and I see no reason why we couldn't use it, but alas first we need to fix the existing assertions to comply with the semantics, i.e. this PR).

ok, lets do the immediate fix first and then the native update in a new PR. @kv2019i is this PR good for the semantic fix ? (if so pls approve).

File: src/audio/eq_iir/eq_iir.c
Comment (kv2019i) [reply src/audio/eq_iir/eq_iir.c:791]: @marc-hb How did you get the empty statement warning? I tried with various empty/NULL definitions of assert (with your PR), but couldn't get a warning. 

@lgirdwood I'd like to resolve this issue before merging. But no other concerns with the PR.

File: src/audio/eq_fir/eq_fir.c
Comment (singalsu) [reply src/audio/eq_fir/eq_fir.c:470]: @marc-hb The PR is #5771

**Comment by marc-hb** [comment]
> @marc-hb Can we sort out the semicolon question and proceed with this PR?

Thank you @singalsu for
- #5771

File: src/audio/eq_iir/eq_iir.c
Comment (marc-hb) [reply src/audio/eq_iir/eq_iir.c:791]: > what is this semicolon here for?

I finally remembered:

```
./scripts/checkpatch.pl -g HEAD
WARNING: braces {} are not necessary for single statement blocks
#67: FILE: src/audio/eq_iir/eq_iir.c:790:
+	if (cmd == COMP_TRIGGER_START || cmd == COMP_TRIGGER_RELEASE) {
 		assert(cd->eq_iir_func);
+	}
```
Anyway #5771 will supersede this hunk.

**Comment by lgirdwood** [comment]
@marc-hb think we are good to go once conflicts fixed. Thanks

**Comment by marc-hb** [comment]
> @marc-hb think we are good to go once conflicts fixed. Thanks

This is depending on #5771 which is small and approved but not merged yet.

**Comment by marc-hb** [comment]
XCC licence glitch in https://sof-ci.01.org/sofpr/PR5758/build269/build/ . Probably temporary because only one build was affected, all others were fine.

**Comment by marc-hb** [comment]
SOFCI TEST

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
